src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,329 @@
+                    }
+
+                    if (readResult.IsCompleted)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
+                    }
+                }
+                finally
+                {
+                    _pipeReader.AdvanceTo(buffer.Start, buffer.End);
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal bool TryParseHeadersToEnd(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            ref long headersLength)
+        {
+            if (buffer.IsSingleSegment)
+            {
+                var didFinishParsing = TryParseHeadersToEndFast(buffer.First.Span,
+                    ref accumulator,
+                    HeadersLengthLimit - headersLength,
+                    out var consumed);
+                headersLength += consumed;
+                buffer = buffer.Slice(consumed);
+                return didFinishParsing;
+            }
+
+            return TryParseHeadersToEndSlow(ref buffer,
+                ref accumulator,
+                ref headersLength);
+        }
+
+        // Fast parsing for single span in ReadOnlySequence
+        private bool TryParseHeadersToEndFast(ReadOnlySpan<byte> span,
+            ref KeyValueAccumulator accumulator,
+            long lengthLimit,
+            out int consumed)
+        {
+            consumed = 0;
+
+            while (span.Length > 0)
+            {
+                // Find the end of the header.
+                var newLine = span.IndexOf(CrlfDelimiter);
+                ReadOnlySpan<byte> line;
+                var foundNewLine = newLine != -1;
+
+                if (foundNewLine)
+                {
+                    line = span.Slice(0, newLine);
+                    span = span.Slice(line.Length + CrlfDelimiter.Length);
+                    consumed += line.Length + CrlfDelimiter.Length;
+                }
+                // We can't know that what is currently read is the end of the header value, that's only the case if this is the final block
+                // If we're not in the final block, then consume nothing
+                else
+                {
+                    // Don't buffer indefinitely
+                    if (span.Length > lengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
+                    }
+                    return false;
+                }
+
+                if (line.Length == 0) // an empty line means it's the end of the headers
+                {
+                    return true;
+                }
+
+                if (line.Length > lengthLimit)
+                {
+                    throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
```

Comment:
I suspect these exceptions are the way they are since I tried sticking to the original behavior, which seems like a limitation from the way it was implemented with the Buffered Read Stream -

I don't think there is in fact a length limit for a specific header, and think it would make sense to use the same error message in all these situations and just use `throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");`

Reply:
And obviously update the tests I was reluctant to update